sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -879,6 +879,36 @@ object TypeCoercion {
           }
         }
         e.withNewChildren(children)
+      case udf: ScalaUDF if udf.inputTypes.nonEmpty =>
+        val children = udf.children.zip(udf.inputTypes).map { case (in, expected) =>
+          implicitCast(in, scalaUDFExpectedTypes(in.dataType, expected)).getOrElse(in)
+        }
+        udf.withNewChildren(children)
+    }
+
+    private def scalaUDFExpectedTypes(input: DataType, expectedType: DataType): DataType = {
+      (input, expectedType) match {
+        // SPARK-26308: avoid casting to an arbitrary precision and scale for decimals. Please note
+        // that precision and scale cannot be inferred properly for a ScalaUDF because, when it is
+        // created, it is not bound to any column. So here the precision and scale of the input
+        // column is used.
+        case (in: DecimalType, _: DecimalType) => in
+        case (ArrayType(dtIn, _), ArrayType(dtExp, nullableExp)) =>
+          ArrayType(scalaUDFExpectedTypes(dtIn, dtExp), nullableExp)
```

Comment:
what's the current behavior of the udf expect `Array[Int]` but the actual column is `ArrayType(IntegerType, nullable = true)`?

Reply:
an `AnalysisException` is thrown